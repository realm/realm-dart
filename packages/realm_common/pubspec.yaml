name: realm_common
description: >-
             Hosts the common code shared between realm, realm_dart and realm_generator packages.
             This package is part of the official Realm Flutter and Realm Dart SDKs.

version: 3.2.0

homepage: https://www.realm.io
repository: https://github.com/realm/realm-dart
issue_tracker: https://github.com/realm/realm-dart/issues

environment:
  sdk: ^3.3.0

dependencies:
  collection: ^1.18.0
  objectid: ^3.0.0
  sane_uuid: ^1.0.0

dev_dependencies:
  lints: ^3.0.0
