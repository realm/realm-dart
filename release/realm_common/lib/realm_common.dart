// Copyright 2021 MongoDB, Inc.
// SPDX-License-Identifier: Apache-2.0

export 'src/realm_common_base.dart';
export 'src/realm_types.dart';

export 'package:objectid/objectid.dart' show ObjectId;
export 'package:sane_uuid/uuid.dart' show Uuid;
