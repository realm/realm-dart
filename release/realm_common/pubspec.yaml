name: realm_common
description: >-
             Hosts the common code shared between realm, realm_dart and realm_generator packages.
             This package is part of the official Realm Flutter and Realm Dart SDKs.

version: 2.0.0-alpha.5

homepage: https://www.realm.io
repository: https://github.com/realm/realm-dart
issue_tracker: https://github.com/realm/realm-dart/issues

environment:
  sdk: ^3.0.0

dependencies:
  collection: ^1.18.0
  objectid: ^3.0.0
  sane_uuid: ^1.0.0-alpha.5

dev_dependencies:
  lints: ^3.0.0
