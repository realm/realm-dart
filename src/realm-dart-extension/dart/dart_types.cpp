////////////////////////////////////////////////////////////////////////////
//
// Copyright 2016 Realm Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////
#pragma clang diagnostic ignored "-Wdocumentation"

#include <cmath>
#include <functional>
#include <map>
#include <string>

#include "dart_api.h"
#include "dart_native_api.h"

#include "js_types.hpp"

#include "dart_types.hpp"

bool ___DartError::use_native_exceptions;

Dart_Handle operator ||(const Dart_Handle& dartResult, ___DartError error) {
	if (Dart_IsError(dartResult)) {
		auto err = Dart_GetError(dartResult);
		//Dart_PropagateError(dartResult);
		if (!___DartError::use_native_exceptions) {
			Dart_ThrowException(dartResult);
		}
		else {
			throw std::runtime_error(err);
		}
		
	}
	return dartResult;
}

namespace realm {
	namespace dartvm {
		namespace flutter {
			std::string FilesDir;
		}

		Dart_Handle DartCoreLibrary;
		Dart_Handle RealmLibrary;
		std::string RealmPackageName;
		Dart_Handle DartDateTimeType;
		
		Dart_Handle RealmCreateDateTimeFunc;
		Dart_Handle RealmDynamicObjectType;
		Dart_Handle RealmTypeStaticPropertiesType;
		Dart_Handle RealmHelpersType;
	}

	namespace util {
		void to_lower(char& c) {
			c = std::tolower(static_cast<unsigned char>(c));
		}

		void to_lowecase(std::string& string)
		{
			std::for_each(string.begin(), string.end(), to_lower);
		}
	}
}

namespace Dart {

	HandleScope::HandleScope() {
		Dart_EnterScope();
	}

	HandleScope::~HandleScope() {
		Dart_ExitScope();
	}

	Dart_Handle GetLibrary(const char* libName) {
		Dart_Handle libraryUrl = Dart_NewStringFromCString(libName) || handleError;
		Dart_Handle library = Dart_LookupLibrary(libraryUrl) || handleError;
		return library;
	}
}

