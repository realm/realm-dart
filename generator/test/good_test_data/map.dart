import 'dart:typed_data';

import 'package:realm_common/realm_common.dart';

//part 'map.g.dart';

@RealmModel()
class _LotsOfMaps {
  late Map<String, _Person?> persons;
  late Map<String, bool> bools;
  late Map<String, DateTime> dateTimes;
  late Map<String, Decimal128> decimals;
  late Map<String, double> doubles;
  late Map<String, int> ints;
  late Map<String, ObjectId> objectIds;
  late Map<String, RealmValue> any;
  late Map<String, String> strings;
  late Map<String, Uint8List> binary;
  late Map<String, Uuid> uuids;
}

@RealmModel()
class _Person {
  late String name;
}
