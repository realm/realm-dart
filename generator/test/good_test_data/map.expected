// **************************************************************************
// RealmObjectGenerator
// **************************************************************************

class LotsOfMaps extends _LotsOfMaps
    with RealmEntity, RealmObjectBase, RealmObject {
  LotsOfMaps({
    Iterable<Person> persons,
    Iterable<bool> bools,
    Iterable<DateTime> dateTimes,
    Iterable<Decimal128> decimals,
    Iterable<double> doubles,
    Iterable<int> ints,
    Iterable<ObjectId> objectIds,
    Iterable<RealmValue> any,
    Iterable<String> strings,
    Iterable<Uint8List> binary,
    Iterable<Uuid> uuids,
  }) {
    RealmObjectBase.set<RealmMap<String, Person>>(
        this, 'persons', RealmMap<String, Person>(persons));
    RealmObjectBase.set<RealmMap<String, bool>>(
        this, 'bools', RealmMap<String, bool>(bools));
    RealmObjectBase.set<RealmMap<String, DateTime>>(
        this, 'dateTimes', RealmMap<String, DateTime>(dateTimes));
    RealmObjectBase.set<RealmMap<String, Decimal128>>(
        this, 'decimals', RealmMap<String, Decimal128>(decimals));
    RealmObjectBase.set<RealmMap<String, double>>(
        this, 'doubles', RealmMap<String, double>(doubles));
    RealmObjectBase.set<RealmMap<String, int>>(
        this, 'ints', RealmMap<String, int>(ints));
    RealmObjectBase.set<RealmMap<String, ObjectId>>(
        this, 'objectIds', RealmMap<String, ObjectId>(objectIds));
    RealmObjectBase.set<RealmMap<String, RealmValue>>(
        this, 'any', RealmMap<String, RealmValue>(any));
    RealmObjectBase.set<RealmMap<String, String>>(
        this, 'strings', RealmMap<String, String>(strings));
    RealmObjectBase.set<RealmMap<String, Uint8List>>(
        this, 'binary', RealmMap<String, Uint8List>(binary));
    RealmObjectBase.set<RealmMap<String, Uuid>>(
        this, 'uuids', RealmMap<String, Uuid>(uuids));
  }

  LotsOfMaps._();

  @override
  RealmMap<String, Person> get persons =>
      RealmObjectBase.get<Person>(this, 'persons') as RealmMap<String, Person>;
  @override
  set persons(covariant RealmMap<String, Person> value) =>
      throw RealmUnsupportedSetError();

  @override
  RealmMap<String, bool> get bools =>
      RealmObjectBase.get<bool>(this, 'bools') as RealmMap<String, bool>;
  @override
  set bools(covariant RealmMap<String, bool> value) =>
      throw RealmUnsupportedSetError();

  @override
  RealmMap<String, DateTime> get dateTimes =>
      RealmObjectBase.get<DateTime>(this, 'dateTimes')
          as RealmMap<String, DateTime>;
  @override
  set dateTimes(covariant RealmMap<String, DateTime> value) =>
      throw RealmUnsupportedSetError();

  @override
  RealmMap<String, Decimal128> get decimals =>
      RealmObjectBase.get<Decimal128>(this, 'decimals')
          as RealmMap<String, Decimal128>;
  @override
  set decimals(covariant RealmMap<String, Decimal128> value) =>
      throw RealmUnsupportedSetError();

  @override
  RealmMap<String, double> get doubles =>
      RealmObjectBase.get<double>(this, 'doubles') as RealmMap<String, double>;
  @override
  set doubles(covariant RealmMap<String, double> value) =>
      throw RealmUnsupportedSetError();

  @override
  RealmMap<String, int> get ints =>
      RealmObjectBase.get<int>(this, 'ints') as RealmMap<String, int>;
  @override
  set ints(covariant RealmMap<String, int> value) =>
      throw RealmUnsupportedSetError();

  @override
  RealmMap<String, ObjectId> get objectIds =>
      RealmObjectBase.get<ObjectId>(this, 'objectIds')
          as RealmMap<String, ObjectId>;
  @override
  set objectIds(covariant RealmMap<String, ObjectId> value) =>
      throw RealmUnsupportedSetError();

  @override
  RealmMap<String, RealmValue> get any =>
      RealmObjectBase.get<RealmValue>(this, 'any')
          as RealmMap<String, RealmValue>;
  @override
  set any(covariant RealmMap<String, RealmValue> value) =>
      throw RealmUnsupportedSetError();

  @override
  RealmMap<String, String> get strings =>
      RealmObjectBase.get<String>(this, 'strings') as RealmMap<String, String>;
  @override
  set strings(covariant RealmMap<String, String> value) =>
      throw RealmUnsupportedSetError();

  @override
  RealmMap<String, Uint8List> get binary =>
      RealmObjectBase.get<Uint8List>(this, 'binary')
          as RealmMap<String, Uint8List>;
  @override
  set binary(covariant RealmMap<String, Uint8List> value) =>
      throw RealmUnsupportedSetError();

  @override
  RealmMap<String, Uuid> get uuids =>
      RealmObjectBase.get<Uuid>(this, 'uuids') as RealmMap<String, Uuid>;
  @override
  set uuids(covariant RealmMap<String, Uuid> value) =>
      throw RealmUnsupportedSetError();

  @override
  Stream<RealmObjectChanges<LotsOfMaps>> get changes =>
      RealmObjectBase.getChanges<LotsOfMaps>(this);

  @override
  LotsOfMaps freeze() => RealmObjectBase.freezeObject<LotsOfMaps>(this);

  static SchemaObject get schema => _schema ??= _initSchema();
  static SchemaObject? _schema;
  static SchemaObject _initSchema() {
    RealmObjectBase.registerFactory(LotsOfMaps._);
    return const SchemaObject(
        ObjectType.realmObject, LotsOfMaps, 'LotsOfMaps', [
      SchemaProperty('persons', RealmPropertyType.object,
          linkTarget: 'Person', collectionType: RealmCollectionType.dictionary),
      SchemaProperty('bools', RealmPropertyType.bool,
          collectionType: RealmCollectionType.dictionary),
      SchemaProperty('dateTimes', RealmPropertyType.timestamp,
          collectionType: RealmCollectionType.dictionary),
      SchemaProperty('decimals', RealmPropertyType.decimal128,
          collectionType: RealmCollectionType.dictionary),
      SchemaProperty('doubles', RealmPropertyType.double,
          collectionType: RealmCollectionType.dictionary),
      SchemaProperty('ints', RealmPropertyType.int,
          collectionType: RealmCollectionType.dictionary),
      SchemaProperty('objectIds', RealmPropertyType.objectid,
          collectionType: RealmCollectionType.dictionary),
      SchemaProperty('any', RealmPropertyType.mixed,
          optional: true, collectionType: RealmCollectionType.dictionary),
      SchemaProperty('strings', RealmPropertyType.string,
          collectionType: RealmCollectionType.dictionary),
      SchemaProperty('binary', RealmPropertyType.binary,
          collectionType: RealmCollectionType.dictionary),
      SchemaProperty('uuids', RealmPropertyType.uuid,
          collectionType: RealmCollectionType.dictionary),
    ]);
  }
}

class Person extends _Person with RealmEntity, RealmObjectBase, RealmObject {
  Person(
    String name,
  ) {
    RealmObjectBase.set(this, 'name', name);
  }

  Person._();

  @override
  String get name => RealmObjectBase.get<String>(this, 'name') as String;
  @override
  set name(String value) => RealmObjectBase.set(this, 'name', value);

  @override
  Stream<RealmObjectChanges<Person>> get changes =>
      RealmObjectBase.getChanges<Person>(this);

  @override
  Person freeze() => RealmObjectBase.freezeObject<Person>(this);

  static SchemaObject get schema => _schema ??= _initSchema();
  static SchemaObject? _schema;
  static SchemaObject _initSchema() {
    RealmObjectBase.registerFactory(Person._);
    return const SchemaObject(ObjectType.realmObject, Person, 'Person', [
      SchemaProperty('name', RealmPropertyType.string),
    ]);
  }
}
